pyop2/codegen/builder.py

```diff
@@ -0,0 +1,816 @@
+        n, i, f = multiindex
+        if layer is not None and self.offset is not None:
+            key = f.extent
+            if key is None:
+                key = 1
+
+            if key not in self.prefetch:
+                bottom_layer, _ = self.layer_bounds
+                offset_extent, = self.offset.shape
+                j = Index(offset_extent)
+                base = Indexed(self.values, (n, j))
+                if f.extent:
+                    k = Index(f.extent)
+                else:
+                    k = Index(1)
+                offset = Sum(Sum(layer, Product(Literal(numpy.int32(-1)), bottom_layer)), k)
+                offset = Product(offset, Indexed(self.offset, (j,)))
+                self.prefetch[key] = Materialise(PackInst(), Sum(base, offset), MultiIndex(k, j))
+
+            return Indexed(self.prefetch[key], (f, i)), (f, i)
+        else:
+            assert f.extent == 1 or f.extent is None
+            base = Indexed(self.values, (n, i))
+            return base, (f, i)
+
+    def indexed_vector(self, n, shape, layer=None):
+        shape = self.shape[1:] + shape
+        if self.interior_horizontal:
+            shape = (2, ) + shape
+        else:
+            shape = (1, ) + shape
+        f, i, j = (Index(e) for e in shape)
+        base, (f, i) = self.indexed((n, i, f), layer=layer)
+        discard = Comparison("<", base, Zero((), self.dtype))
+        if self.vector_bc is not None:
+            # Exposition:
+            # Vector-index bcs are encoded in the high bits of the map.
+            # The incoming value is:
+            # input := ~(row + sum_i 2**(nbit - i))
+            # Where i are the components to zero
+            # The actual row is then:
+            # row := (~input) & (~(sum_{k<3} 2**(nbit - k)))
+            # And the high bits that are non-zero tell us which
+            # values to mask out.
+            nbits = Literal(self.dtype.type(self.dtype.itemsize*8 - 2))
+            mask = Literal(self.dtype.type(sum(2**(nbits.value - i) for i in range(3))))
+            flipped = BitwiseNot(base)
+            base = Conditional(discard,
+                               BitwiseAnd(flipped, BitwiseNot(mask)),
+                               base)
+            expr = LogicalNot(BitwiseAnd(flipped, mask))
+            expr = LogicalOr(expr,
+                             BitwiseAnd(flipped,
+                                        BitShift("<<", Literal(self.dtype.type(1)),
+                                                 Sum(nbits,
+                                                     Product(Literal(self.dtype.type(-1)), j)))))
+            discard = LogicalAnd(discard, expr)
+
+        init = Conditional(discard, Literal(self.dtype.type(-1)), Sum(Product(base, Literal(numpy.int32(j.extent))), j))
+        pack = Materialise(PackInst(), init, MultiIndex(f, i, j))
+        multiindex = tuple(Index(e) for e in pack.shape)
+        return Indexed(pack, multiindex), multiindex
```

Comment:
Sounds good, this simplifies the initialisers as well.
Though do we want to make this PR bigger still or should we do this as a separate PR (since this touches firedrake too)?